### ISSUE TITLE
When SlashCommand options are defined within the decorator rather than command parameters, an AttributeError exception will be thrown upon invoking the command.

### ISSUE THREAD

**Issue by tcrory** [Pycord-Development/pycord#555]
### Summary

When a slash command is created, the user is given the option of defining `Option`s using the command parameters or by passing in a list through the decorator. Passing the list through the decorator bypasses the `_parse_options()` method, which is the only place where the `_parameter_name` attribute gets set for an `Option`. This attribute is then referenced within the `_invoke()` method, which will throw if the options were not defined using the `_parse_options()` method.

### Reproduction Steps

Define a slash command where the options are passed into the `@slash_command()` decorator instead of through type hinting on the command method.

### Expected Results

The command would be processed, and the command would be successfully invoked when received.

### Actual Results

The invocation of the command references the `Option._parameter_name` attribute, which only gets set by the `_parse_options()` method when the user does not provide the options through the `slash_command()` decorator. If the attribute has not been set, an exception will be thrown:
`AttributeError: 'Option' object has no attribute '_parameter_name'`

```
Ignoring exception in command <discord.commands.SlashCommand name=list>:
Traceback (most recent call last):
  File "C:\Projects\support-bot\env\lib\site-packages\discord\commands\c
    ret = await coro(arg)
  File "C:\Projects\support-bot\env\lib\site-packages\discord\commands\c
    kwargs[op._parameter_name] = arg
AttributeError: 'Option' object has no attribute '_parameter_name'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Projects\support-bot\env\lib\site-packages\discord\bot.py", l
    await ctx.command.invoke(ctx)
  File "C:\Projects\support-bot\env\lib\site-packages\discord\commands\c
    await injected(ctx)
  File "C:\Projects\support-bot\env\lib\site-packages\discord\commands\c
    raise ApplicationCommandInvokeError(exc) from exc
discord.commands.errors.ApplicationCommandInvokeError: Application Command raise
Ignoring exception in command <discord.commands.SlashCommand name=list>:
Traceback (most recent call last):
  File "C:\Projects\support-bot\env\lib\site-packages\discord\commands\c
    ret = await coro(arg)
  File "C:\Projects\support-bot\env\lib\site-packages\discord\commands\c
    if o._parameter_name not in kwargs:
AttributeError: 'Option' object has no attribute '_parameter_name'
ine 444, in process_application_commands
    await ctx.command.invoke(ctx)
  File "C:\Projects\support-bot\env\lib\site-packages\discord\commands\ccommands\commands.py", line 134, in invoke
    await injected(ctx)
  File "C:\Projects\support-bot\env\lib\site-packages\discord\commands\ccommands\commands.py", line 91, in wrapped
    raise ApplicationCommandInvokeError(exc) from exc
discord.commands.errors.ApplicationCommandInvokeError: Application Command raisemand raised an exception: AttributeError: 'Option' object has no attriameter_nambute '_parameter_name'
```

### Intents

Default plus `members = true`. (Presences and typing both false.)

### System Information

```
python -m discord -v
- Python v3.8.5-final
- py-cord v2.0.0-alpha
    - py-cord pkg_resources: v2.0.0a4509+g8a7b57bb
- aiohttp v3.7.4.post0
- system info: Windows 10 10.0.19041
```

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

I started to implement a fix for this with the intent of creating a PR, but I realized that without understanding the purpose for using Option._parameter_name, I'm not entirely sure what the correct course of action would be. The two options I found myself considering were to reference `Option.name` instead of `Option._parameter_name` when invoking a command, or alternatively setting `_parameter_name` whenever a list of options are passed in through the decorator (bypassing the `_parse_options()` method).

This second option made me think that it should probably be part of the `Option` constructor, at which point I realized that the correct solution likely requires more context on what the original intent was for adding the `_parameter_name` attribute.

**Comment by tcrory** [comment]
While searching for context, I found that the `Option._parameter_name` attribute was added in #374, which was a fix for #106.

**Comment by EmreTech** [comment]
After looking through the code, `Option._parameter_name` seems to be assigned based on the name given to an Option type hint when you define a `slash_command`.

Here's an example if that made no sense:
```python
@slash_command()
async def ping(ctx, person: Option(discord.User)):
                 await ctx.respond("Hi")
```
The `_parameter_name` would be set to "person" in this case

**Comment by PythonIntermediateCoder** [comment]
I'm not sure if It's a bug, can you please show me the reproductible code?

**Comment by EmreTech** [comment]
> I'm not sure if It's a bug, can you please show me the reproductible code?

#680 (which is a duplicate) has some reproducible code

**Comment by Luc1412** [comment]
This bug was unintentionally introduced by myself https://github.com/Pycord-Development/pycord/pull/471. I wasn't aware that pycord already supported different naming for parameter and option name. 

A possible fix would be using a Dict like: 
```py
options={
    "param_name": discord.Option(...)
}
```

**Comment by EmreTech** [comment]
> This bug was unintentionally introduced by myself https://github.com/Pycord-Development/pycord/pull/471. I wasn't aware that pycord already supported different naming for parameter and option name. 
> 
> A possible fix would be using a Dict like: 
> ```py
> options={
>     "param_name": discord.Option(...)
> }
> ```
> 

Actually, funny enough, someone on the Pycord Discord Server had the same idea. It's a possibility, but it's redundant because you pass in everything you list in the Option kwarg into the function definition in order.

**Comment by plun1331** [comment]
Another fix could be basing it on the order of the function parameters

**Comment by EmreTech** [comment]
> Another fix could be basing it on the order of the function parameters

This is implemented in #729

**Comment by EmreTech** [comment]
The PR tackling this issue, #729, has been merged. This issue shall be closed now.
